EXTRACCION, TRANSFORMACION Y CARGA

In [86]:
#iniciamos importando las librerias
import pandas as pd
import warnings
warnings.filterwarnings("ignore") 

PROCESO DE EXTRACCION DE DATOS


In [ ]:
#Iniciamos carganado los datasets desde el archivo 'homicidios.xlsx' que contienen dos hojas
#llamadas 'VICTIMAS', 'HECHOS' y procedemos a guardarlas en dos dataframes diferentes

In [87]:
hechos_df = pd.read_excel("homicidios.xlsx", sheet_name= "HECHOS")

In [88]:
victimas_df = pd.read_excel("homicidios.xlsx", sheet_name= "VICTIMAS")

ENTENDIENDO LOS DATOS

In [89]:
#Investigamos informacion en general de el dataframe 'hechos_df'

In [90]:
hechos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

BUSCAMOS VALORES NULOS

In [91]:
hechos_df.isnull().sum()

ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64

lo que podemos analisar es que el dataframe esta bastante organizado, encontramos algunos valores nulos en algunas columnas  como los son: 'Calle', 'Altura', 'Cruce' y 'Direccion Normalizada' en este caso ya que es un estudio estadistico no tiene para nosotros interes de alto valor asi que procedemos a eliminarlas

In [92]:
#Eliminamos las columnas  con valores nulos
hechos_sin_nulos_df = hechos_df.dropna(axis=1)

In [93]:
#verificamos nuevame la informacion general
hechos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

COLUMNAS 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', Y 'HH'

In [ ]:
#Estas columnas pueden ser un poco redundantes, pero no es necesario eliminarlas ya que no tienen valores nulos que puedan de pronto entorpecer las consultas
#asi que decido mantenerlas, solo queda verificar los tipos  de datos.

In [94]:
#nos damos de cuenta que hay al  menos un registro con 'SD', en la  columna 'HORA'
hechos_sd = hechos_sin_nulos_df[hechos_sin_nulos_df['HORA'] == 'SD'][['ID', 'FECHA', 'HORA', 'HH']]

In [95]:
#verificamos
hechos_sd

ID      FECHA HORA  HH
518  2019-0103 2019-12-18   SD  SD

In [96]:
#procedemos a cambiar el valor de 'SD' por '0' para luego poder transformar el tipo de dato de estas columnas
hechos_sin_nulos_df ['HORA'][518] = 0
hechos_sin_nulos_df ['HH'][518] = 0

Columnas LUGAR_DEL_HECHO y XY (CABA)

In [ ]:
#estas columnas tampoco nos van a servir para el analisis 

In [97]:
#eliminamos las columnas en especifica
hechos_sin_nulos_df = hechos_sin_nulos_df.drop(columns=['LUGAR_DEL_HECHO', 'XY (CABA)'])

In [98]:
#Verificamos el dataframe luego eliminar los valores nulos
hechos_sin_nulos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   ID             696 non-null    object        
 1   N_VICTIMAS     696 non-null    int64         
 2   FECHA          696 non-null    datetime64[ns]
 3   AAAA           696 non-null    int64         
 4   MM             696 non-null    int64         
 5   DD             696 non-null    int64         
 6   HORA           696 non-null    object        
 7   HH             696 non-null    object        
 8   TIPO_DE_CALLE  696 non-null    object        
 9   COMUNA         696 non-null    int64         
 10  pos x          696 non-null    object        
 11  pos y          696 non-null    object        
 12  PARTICIPANTES  696 non-null    object        
 13  VICTIMA        696 non-null    object        
 14  ACUSADO        696 non-null    object        
dtypes: datetime64[ns](1), i

Columnas VICTIMA y ACUSADO

In [ ]:
#estas columnas no tienen valores nulos pero si tiene valores duplicados

In [99]:
hechos_sin_nulos_df['ACUSADO'].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

In [100]:
hechos_sin_nulos_df['VICTIMA'].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

verificamos que la columna 'pos x' y 'pos y' tienen diferentes datos mal cargados como solo un punto, pero no tiene datos nulos, por ende vamos a corregir, el problema es que no identificamos donde estan , le damos una ubicacion fuera de CABA unificada.

In [102]:
#seleciconamos las filas 'pos x' es igual a '.'
actualizar_filas = hechos_sin_nulos_df[hechos_sin_nulos_df['pos x'] == '.']

In [103]:
#Actualizamos las filas seleccionadas con los valores deseados
hechos_sin_nulos_df.loc[actualizar_filas.index, 'pos x'] = -58.526752
hechos_sin_nulos_df.loc[actualizar_filas.index, 'pos y'] = -34.546054

VERIFICAMOS EL TIPO DE LAS VARIABLES

In [104]:
hechos_sin_nulos_df.dtypes

ID                       object
N_VICTIMAS                int64
FECHA            datetime64[ns]
AAAA                      int64
MM                        int64
DD                        int64
HORA                     object
HH                       object
TIPO_DE_CALLE            object
COMUNA                    int64
pos x                    object
pos y                    object
PARTICIPANTES            object
VICTIMA                  object
ACUSADO                  object
dtype: object

In [105]:
hechos_sin_nulos_df['ID'] = hechos_sin_nulos_df['ID'].astype(str)
hechos_sin_nulos_df['HORA'] = hechos_sin_nulos_df['HORA'].astype(str)
hechos_sin_nulos_df['HH'] = hechos_sin_nulos_df['HH']. astype(int)
hechos_sin_nulos_df['TIPO_DE_CALLE'] = hechos_sin_nulos_df['TIPO_DE_CALLE']. astype(str)
hechos_sin_nulos_df['pos x'] = hechos_sin_nulos_df['pos x']. astype(float)
hechos_sin_nulos_df['pos y'] = hechos_sin_nulos_df['pos y']. astype(float)
hechos_sin_nulos_df['PARTICIPANTES'] = hechos_sin_nulos_df['PARTICIPANTES']. astype(str)
hechos_sin_nulos_df['VICTIMA'] = hechos_sin_nulos_df['VICTIMA']. astype(str)
hechos_sin_nulos_df['ACUSADO'] = hechos_sin_nulos_df['ACUSADO']. astype(str)

Le asiganmos nombres mas legibles para que cada una de los nombres de las columnas sean mas faciles de trabajar

In [106]:
hechos_final_df = hechos_sin_nulos_df.rename(columns={'N_VICTIMAS': 'Numero de víctimas',
                                'FECHA': 'Fecha',
                                'AAAA':'Año',
                                'MM':'Mes',
                                'DD':'Día',
                                'HORA': 'Hora',
                                'HH':'Hora entera',
                                'TIPO_DE_CALLE': 'Tipo de Calle',
                                'COMUNA': 'Comuna',
                                'pos x': 'Pos x',
                                'pos y': 'Pos y',
                                'PARTICIPANTES': 'Participantes',                      
                                'VICTIMA': 'Víctima',
                                'ACUSADO': 'Acusado'})

In [107]:
#verificamos
hechos_final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   ID                  696 non-null    object        
 1   Numero de víctimas  696 non-null    int64         
 2   Fecha               696 non-null    datetime64[ns]
 3   Año                 696 non-null    int64         
 4   Mes                 696 non-null    int64         
 5   Día                 696 non-null    int64         
 6   Hora                696 non-null    object        
 7   Hora entera         696 non-null    int32         
 8   Tipo de Calle       696 non-null    object        
 9   Comuna              696 non-null    int64         
 10  Pos x               696 non-null    float64       
 11  Pos y               696 non-null    float64       
 12  Participantes       696 non-null    object        
 13  Víctima             696 non-null    object        

In [109]:
#Exportamos a CSV para continuear con el trabajo
hechos_final_df.to_csv('C:/Users/CRISTIAN/Desktop/proyecto final # 2/CSV/hechos_homicidios.csv')

Analisis exploratorio y transformacion de los datos para el dataframe de VICTIMAS

In [ ]:
#Investigamos informacion en general de el dataframe 'victimas_df'

In [110]:
#
victimas_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


In [111]:
victimas_df.head(10)

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
0  2016-0001 2016-01-01  2016   1   1             CONDUCTOR    MOTO   
1  2016-0002 2016-01-02  2016   1   2             CONDUCTOR    AUTO   
2  2016-0003 2016-01-03  2016   1   3             CONDUCTOR    MOTO   
3  2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
4  2016-0005 2016-01-21  2016   1  21             CONDUCTOR    MOTO   
5  2016-0008 2016-01-24  2016   1  24             CONDUCTOR    MOTO   
6  2016-0009 2016-01-24  2016   1  24  PASAJERO_ACOMPAÑANTE    MOTO   
7  2016-0010 2016-01-29  2016   1  29             CONDUCTOR    MOTO   
8  2016-0012 2016-02-08  2016   2   8             CONDUCTOR    MOTO   
9  2016-0013 2016-02-10  2016   2  10                PEATON  PEATON   

        SEXO EDAD  FECHA_FALLECIMIENTO  
0  MASCULINO   19  2016-01-01 00:00:00  
1  MASCULINO   70  2016-01-02 00:00:00  
2  MASCULINO   30  2016-01-03 00:00:00  
3  MASCULINO   18                   SD  
4  MASCULINO   29  2016-02-01 00:00:00  
5  MASCULINO   30  2016-01-24 00:00:00  
6  MASCULINO   29  2016-01-26 00:00:00  
7  MASCULINO   18  2016-01-29 00:00:00  
8  MASCULINO   22  2016-02-08 00:00:00  
9  MASCULINO   16  2016-02-10 00:00:00

In [112]:
hechos_final_df.head(10)

ID  Numero de víctimas      Fecha   Año  Mes  Día      Hora  \
0  2016-0001                   1 2016-01-01  2016    1    1  04:00:00   
1  2016-0002                   1 2016-01-02  2016    1    2  01:15:00   
2  2016-0003                   1 2016-01-03  2016    1    3  07:00:00   
3  2016-0004                   1 2016-01-10  2016    1   10  00:00:00   
4  2016-0005                   1 2016-01-21  2016    1   21  05:20:00   
5  2016-0008                   1 2016-01-24  2016    1   24  18:30:00   
6  2016-0009                   1 2016-01-24  2016    1   24  19:10:00   
7  2016-0010                   1 2016-01-29  2016    1   29  15:20:00   
8  2016-0012                   1 2016-02-08  2016    2    8  01:20:00   
9  2016-0013                   1 2016-02-10  2016    2   10  11:30:00   

   Hora entera Tipo de Calle  Comuna      Pos x      Pos y     Participantes  \
0            4       AVENIDA       8 -58.475340 -34.687570         MOTO-AUTO   
1            1      GRAL PAZ       9 -58.508775 -34.669777    AUTO-PASAJEROS   
2            7       AVENIDA       1 -58.390403 -34.631894         MOTO-AUTO   
3            0       AVENIDA       8 -58.465039 -34.680930           MOTO-SD   
4            5       AVENIDA       1 -58.387183 -34.622466    MOTO-PASAJEROS   
5           18       AVENIDA       8 -58.444513 -34.684759  MOTO-OBJETO FIJO   
6           19         CALLE      11 -58.500959 -34.608254         MOTO-AUTO   
7           15      GRAL PAZ       9 -58.508775 -34.669777         MOTO-AUTO   
8            1       AVENIDA       1 -58.380486 -34.613039       MOTO-CARGAS   
9           11       AVENIDA       1 -58.391149 -34.624774       PEATON-AUTO   

  Víctima      Acusado  
0    MOTO         AUTO  
1    AUTO    PASAJEROS  
2    MOTO         AUTO  
3    MOTO           SD  
4    MOTO    PASAJEROS  
5    MOTO  OBJETO FIJO  
6    MOTO         AUTO  
7    MOTO         AUTO  
8    MOTO       CARGAS  
9  PEATON         AUTO

Como podemos darnos de cuenta en los dos dataframe coinciden los datos de la columna 'ID', la unica diferencia es que el dataframe 'HECHOS' tiene 696 filas, mientras que el dataframe 'VICTIMAS' tiene 717 filas ya que vairan en que tienen 2 o mas victimas.

REVISAMOS DATOS NULOS

In [113]:
victimas_df.isnull().sum()

ID_hecho               0
FECHA                  0
AAAA                   0
MM                     0
DD                     0
ROL                    0
VICTIMA                0
SEXO                   0
EDAD                   0
FECHA_FALLECIMIENTO    0
dtype: int64

REVISAMOS DATOS DUPLICADOS

In [ ]:
#A diferencia de el dataframe anterior, no utilizamos el 'ID' para verificar los datos duplicados, pues sabremos que habra duplicados
#en los casos de que un accidente hay tenido varias victimas.

In [114]:
victimas_df[victimas_df.duplicated()]

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD, FECHA_FALLECIMIENTO]
Index: []

al revisar el dataset nos damos de cuenta que en algunas  columnas encontramos el valor 'SD', que significa 'sin dato' revisaremos cada una de las columnas para determinar si es posible alguna imputacion en estos espacios en sin dato.

In [115]:
victimas_df[victimas_df['ID_hecho'] == 'SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD, FECHA_FALLECIMIENTO]
Index: []

In [116]:
victimas_df[victimas_df['FECHA'] == 'SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD, FECHA_FALLECIMIENTO]
Index: []

In [117]:
victimas_df[victimas_df['AAAA'] == 'SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD, FECHA_FALLECIMIENTO]
Index: []

In [118]:
victimas_df[victimas_df['MM'] == 'SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD, FECHA_FALLECIMIENTO]
Index: []

In [119]:
victimas_df[victimas_df['DD'] == 'SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD, FECHA_FALLECIMIENTO]
Index: []

Como podemos darnos cuenta las columnas correspondientes al ID y las relacionadas con la fecha del accidentes estan completas.

In [120]:
victimas_df[victimas_df['FECHA_FALLECIMIENTO'] == 'SD']

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
3    2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
16   2016-0022 2016-02-21  2016   2  21  PASAJERO_ACOMPAÑANTE    MOTO   
19   2016-0027 2016-02-28  2016   2  28  PASAJERO_ACOMPAÑANTE    AUTO   
22   2016-0031 2016-03-08  2016   3   8             CONDUCTOR    MOTO   
33   2016-0045 2016-04-11  2016   4  11             CONDUCTOR    MOTO   
..         ...        ...   ...  ..  ..                   ...     ...   
221  2017-0089 2017-07-13  2017   7  13                    SD      SD   
242  2017-0112 2017-09-10  2017   9  10  PASAJERO_ACOMPAÑANTE    AUTO   
246  2017-0115 2017-09-19  2017   9  19             CONDUCTOR    MOTO   
253  2017-0126 2017-10-14  2017  10  14  PASAJERO_ACOMPAÑANTE    AUTO   
280  2017-0155 2017-12-12  2017  12  12                    SD      SD   

          SEXO EDAD FECHA_FALLECIMIENTO  
3    MASCULINO   18                  SD  
16   MASCULINO   41                  SD  
19   MASCULINO   34                  SD  
22   MASCULINO   21                  SD  
33   MASCULINO   SD                  SD  
..         ...  ...                 ...  
221  MASCULINO   23                  SD  
242  MASCULINO    1                  SD  
246  MASCULINO   34                  SD  
253  MASCULINO   39                  SD  
280  MASCULINO   77                  SD  

[68 rows x 10 columns]

nos damos de cuenta que tenemos 68 registros sin fecha de fallecimiento. procedemos a eliminar las columna 'FECHA_FALLECIMIENTO' ya que como este numero de datos 'SD' significan casi un 10% de las filas, y ser mas grandes que el kpi buscado.

In [122]:
victimas_df = victimas_df.drop(columns=['FECHA_FALLECIMIENTO'])

In [123]:
victimas_df[victimas_df['ROL'] == 'SD']

ID_hecho      FECHA  AAAA  MM  DD ROL VICTIMA       SEXO EDAD
36   2016-0049 2016-04-17  2016   4  17  SD      SD         SD   SD
39   2016-0052 2016-04-20  2016   4  20  SD    MOTO         SD   SD
63   2016-0085 2016-06-29  2016   6  29  SD    MOTO  MASCULINO   SD
77   2016-0101 2016-08-07  2016   8   7  SD      SD  MASCULINO   67
89   2016-0115 2016-09-02  2016   9   2  SD      SD  MASCULINO   SD
141  2016-0174 2016-12-27  2016  12  27  SD      SD         SD   SD
167  2017-0029 2017-03-07  2017   3   7  SD      SD  MASCULINO   34
208  2017-0074 2017-06-04  2017   6   4  SD      SD  MASCULINO   70
221  2017-0089 2017-07-13  2017   7  13  SD      SD  MASCULINO   23
280  2017-0155 2017-12-12  2017  12  12  SD      SD  MASCULINO   77
390  2018-0100 2018-09-20  2018   9  20  SD    MOTO   FEMENINO   33

In [124]:
victimas_df[victimas_df['VICTIMA'] == 'SD']

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
36   2016-0049 2016-04-17  2016   4  17                    SD      SD   
77   2016-0101 2016-08-07  2016   8   7                    SD      SD   
89   2016-0115 2016-09-02  2016   9   2                    SD      SD   
93   2016-0119 2016-09-04  2016   9   4  PASAJERO_ACOMPAÑANTE      SD   
141  2016-0174 2016-12-27  2016  12  27                    SD      SD   
167  2017-0029 2017-03-07  2017   3   7                    SD      SD   
208  2017-0074 2017-06-04  2017   6   4                    SD      SD   
221  2017-0089 2017-07-13  2017   7  13                    SD      SD   
280  2017-0155 2017-12-12  2017  12  12                    SD      SD   

          SEXO EDAD  
36          SD   SD  
77   MASCULINO   67  
89   MASCULINO   SD  
93    FEMENINO   SD  
141         SD   SD  
167  MASCULINO   34  
208  MASCULINO   70  
221  MASCULINO   23  
280  MASCULINO   77

In [125]:
victimas_df[victimas_df['SEXO'] == 'SD']

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA SEXO EDAD
36   2016-0049 2016-04-17  2016   4  17         SD      SD   SD   SD
39   2016-0052 2016-04-20  2016   4  20         SD    MOTO   SD   SD
108  2016-0136 2016-10-25  2016  10  25  CONDUCTOR    MOTO   SD   SD
121  2016-0151 2016-11-18  2016  11  18     PEATON  PEATON   SD   SD
138  2016-0171 2016-12-25  2016  12  25  CONDUCTOR    MOTO   SD   SD
141  2016-0174 2016-12-27  2016  12  27         SD      SD   SD   SD


En las columnas SEXO, ROL y VÍCTIMA, los valores faltantes son mínimos: 0.8%, 1.5% y 1.3%, respectivamente. Por lo tanto, es factible llenar los espacios vacíos de SEXO y ROL con el valor más común en cada columna sin afectar de manera significativa el análisis estadístico.

Optaré por no modificar los datos faltantes en la columna VÍCTIMA para evitar inconsistencias cuando se combine este dataframe con el de Hechos. Posteriormente revisaré nuevamente este campo al fusionar los datos.

In [126]:
valor_frecuente = victimas_df['SEXO'].mode().iloc[0]
print(f'El sexo más frecuente es: {valor_frecuente}')

El sexo más frecuente es: MASCULINO


In [127]:
victimas_df['SEXO'][victimas_df['SEXO']=='SD'] = 'MASCULINO'

In [128]:
victimas_df[victimas_df['SEXO']=='SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD]
Index: []

In [137]:
valor_frecuente = victimas_df['ROL'].mode().iloc[0]
print(f'El rol más frecuente es: {valor_frecuente}')

El rol más frecuente es: CONDUCTOR


In [134]:
victimas_df['ROL'][victimas_df['ROL']=='SD'] = 'CONDUCTOR'

En esta parte del analisis lo que hicimos fue remplazar los registros sin dato por cada una de las frecuencias mas altas de cada variable.

In [135]:
victimas_df[victimas_df['ROL']=='SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD]
Index: []

In [136]:
victimas_df[victimas_df['EDAD'] == 'SD'].shape[0]

53


La columna de EDAD tiene un número considerable de registros incompletos, representando aproximadamente el 6.3% del total. Utilizar 'SD' como marcador para estos datos faltantes podría causar inconvenientes con el tipo de dato en análisis posteriores, y asignar un valor de 0 distorsionaría cualquier cálculo que involucre la edad. Por lo tanto, una solución efectiva es calcular el promedio de edad según el sexo de las víctimas y usar estos valores promedio para rellenar los espacios vacíos.

In [138]:
# Remplzamos el 'SD' por Na para poder calcular el promedio
victimas_df['EDAD'] = victimas_df['EDAD'].replace('SD', pd.NA)

In [149]:
promedio_genero = victimas_df.groupby('SEXO')['EDAD'].mean()
print(f'La edad promedio de FEMENINO es {round(promedio_genero["FEMENINO"])} y de MASCULINO es {round(promedio_genero["MASCULINO"])}')

La edad promedio de FEMENINO es 51 y de MASCULINO es 39


In [150]:
#llenamos los valores NaN en la columna 'EDAD' utilizando el promedio correspondiente de cada genero
victimas_df['EDAD'] = victimas_df.apply(lambda row: promedio_genero[row['SEXO']] if pd.isna(row['EDAD']) else row ['EDAD'], axis=1)
#convertimos el tipo de dato de esta columna 'EDAD' a 'int'
victimas_df['EDAD'] = victimas_df['EDAD'].astype(int) 

In [151]:
victimas_df[victimas_df['EDAD']=='SD']

Empty DataFrame
Columns: [ID_hecho, FECHA, AAAA, MM, DD, ROL, VICTIMA, SEXO, EDAD]
Index: []

In [152]:
#y verificamos el  dataframe
victimas_df.head(25)

ID_hecho      FECHA  AAAA  MM  DD                   ROL VICTIMA  \
0   2016-0001 2016-01-01  2016   1   1             CONDUCTOR    MOTO   
1   2016-0002 2016-01-02  2016   1   2             CONDUCTOR    AUTO   
2   2016-0003 2016-01-03  2016   1   3             CONDUCTOR    MOTO   
3   2016-0004 2016-01-10  2016   1  10             CONDUCTOR    MOTO   
4   2016-0005 2016-01-21  2016   1  21             CONDUCTOR    MOTO   
5   2016-0008 2016-01-24  2016   1  24             CONDUCTOR    MOTO   
6   2016-0009 2016-01-24  2016   1  24  PASAJERO_ACOMPAÑANTE    MOTO   
7   2016-0010 2016-01-29  2016   1  29             CONDUCTOR    MOTO   
8   2016-0012 2016-02-08  2016   2   8             CONDUCTOR    MOTO   
9   2016-0013 2016-02-10  2016   2  10                PEATON  PEATON   
10  2016-0015 2016-02-14  2016   2  14                PEATON  PEATON   
11  2016-0016 2016-02-14  2016   2  14  PASAJERO_ACOMPAÑANTE    MOTO   
12  2016-0017 2016-02-15  2016   2  15                PEATON  PEATON   
13  2016-0019 2016-02-17  2016   2  17                PEATON  PEATON   
14  2016-0020 2016-02-17  2016   2  17                PEATON  PEATON   
15  2016-0021 2016-02-17  2016   2  17             CONDUCTOR    AUTO   
16  2016-0022 2016-02-21  2016   2  21  PASAJERO_ACOMPAÑANTE    MOTO   
17  2016-0025 2016-02-28  2016   2  28  PASAJERO_ACOMPAÑANTE    AUTO   
18  2016-0026 2016-02-28  2016   2  28             CONDUCTOR    AUTO   
19  2016-0027 2016-02-28  2016   2  28  PASAJERO_ACOMPAÑANTE    AUTO   
20  2016-0028 2016-03-02  2016   3   2                PEATON  PEATON   
21  2016-0029 2016-03-04  2016   3   4                PEATON  PEATON   
22  2016-0031 2016-03-08  2016   3   8             CONDUCTOR    MOTO   
23  2016-0034 2016-03-12  2016   3  12                PEATON  PEATON   
24  2016-0035 2016-03-13  2016   3  13             CONDUCTOR    MOTO   

         SEXO  EDAD  
0   MASCULINO    19  
1   MASCULINO    70  
2   MASCULINO    30  
3   MASCULINO    18  
4   MASCULINO    29  
5   MASCULINO    30  
6   MASCULINO    29  
7   MASCULINO    18  
8   MASCULINO    22  
9   MASCULINO    16  
10   FEMENINO    16  
11   FEMENINO    19  
12  MASCULINO    59  
13   FEMENINO    65  
14  MASCULINO    59  
15  MASCULINO    34  
16  MASCULINO    41  
17  MASCULINO    34  
18  MASCULINO    30  
19  MASCULINO    34  
20  MASCULINO    50  
21  MASCULINO    38  
22  MASCULINO    21  
23  MASCULINO    52  
24  MASCULINO    34

verificacion y correccion de los tipos de cada variable

In [153]:
victimas_df.dtypes

ID_hecho            object
FECHA       datetime64[ns]
AAAA                 int64
MM                   int64
DD                   int64
ROL                 object
VICTIMA             object
SEXO                object
EDAD                 int32
dtype: object

In [154]:
victimas_df['ID_hecho'] = victimas_df['ID_hecho'].astype(str)
victimas_df['ROL']=victimas_df['ROL'].astype(str)
victimas_df['VICTIMA']=victimas_df['VICTIMA'].astype(str)
victimas_df['SEXO']=victimas_df['SEXO'].astype(str)

Cambiamos el nombre de las columnas a un nombre mas legible para trabajar

In [155]:
victimas_df = victimas_df.rename(columns={'ID_hecho': 'ID',
                                'FECHA': 'Fecha',
                                'AAAA':'Año',
                                'MM':'Mes',
                                'DD':'Día',
                                'ROL': 'Rol',                               
                                'VICTIMA': 'Víctima',
                                'SEXO': 'Sexo',
                                'EDAD':'Edad'                                
                                })

In [156]:
#verificamos los cambios
victimas_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   ID       717 non-null    object        
 1   Fecha    717 non-null    datetime64[ns]
 2   Año      717 non-null    int64         
 3   Mes      717 non-null    int64         
 4   Día      717 non-null    int64         
 5   Rol      717 non-null    object        
 6   Víctima  717 non-null    object        
 7   Sexo     717 non-null    object        
 8   Edad     717 non-null    int32         
dtypes: datetime64[ns](1), int32(1), int64(3), object(4)
memory usage: 47.7+ KB


In [157]:
#Exportamos a CSV para continuear con el trabajo
victimas_df.to_csv('C:/Users/CRISTIAN/Desktop/proyecto final # 2/CSV/victimas_homicidios.csv')

PREPARAMOS LOS DATASETS PARA EL SIGUIENTE PROCESO QUE ES EL EDA

In [ ]:
#cargamos los datasets ya limpios de los archivos csv

In [158]:
hoja_hechos_df = pd.read_csv('CSV/hechos_homicidios.csv', index_col=0)
hoja_victimas_df = pd.read_csv('CSV/victimas_homicidios.csv', index_col=0)

In [159]:
hoja_hechos_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 696 entries, 0 to 695
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  696 non-null    object 
 1   Numero de víctimas  696 non-null    int64  
 2   Fecha               696 non-null    object 
 3   Año                 696 non-null    int64  
 4   Mes                 696 non-null    int64  
 5   Día                 696 non-null    int64  
 6   Hora                696 non-null    object 
 7   Hora entera         696 non-null    int64  
 8   Tipo de Calle       696 non-null    object 
 9   Comuna              696 non-null    int64  
 10  Pos x               696 non-null    float64
 11  Pos y               696 non-null    float64
 12  Participantes       696 non-null    object 
 13  Víctima             696 non-null    object 
 14  Acusado             696 non-null    object 
dtypes: float64(2), int64(6), object(7)
memory usage: 87.0+ KB


In [160]:
hoja_victimas_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 717 entries, 0 to 716
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ID       717 non-null    object
 1   Fecha    717 non-null    object
 2   Año      717 non-null    int64 
 3   Mes      717 non-null    int64 
 4   Día      717 non-null    int64 
 5   Rol      717 non-null    object
 6   Víctima  717 non-null    object
 7   Sexo     717 non-null    object
 8   Edad     717 non-null    int64 
dtypes: int64(4), object(5)
memory usage: 56.0+ KB


como podemos darnos cuenta tenemos columnas repetidas y procedo a eliminar algunas para no tener inconvenientes con la union de los dataframes

In [162]:
hoja_victimas_df = hoja_victimas_df.drop(columns=['Fecha', 'Año', 'Mes', 'Día'])

In [164]:
hoja_hechos_df = hoja_hechos_df.drop(columns=['Víctima'])

realizamos el metodo 'merge' entre los dos dataframes para obtener un solo dataframe y poder proseguir a trabajar con el EDA y la confeccion de el Dasboard

In [165]:
homicidos_df = pd.merge(hoja_hechos_df, hoja_victimas_df, on='ID', how='inner')
homicidos_df

ID  Numero de víctimas       Fecha   Año  Mes  Día      Hora  \
0    2016-0001                   1  2016-01-01  2016    1    1  04:00:00   
1    2016-0002                   1  2016-01-02  2016    1    2  01:15:00   
2    2016-0003                   1  2016-01-03  2016    1    3  07:00:00   
3    2016-0004                   1  2016-01-10  2016    1   10  00:00:00   
4    2016-0005                   1  2016-01-21  2016    1   21  05:20:00   
..         ...                 ...         ...   ...  ...  ...       ...   
712  2021-0093                   1  2021-12-13  2021   12   13  17:10:00   
713  2021-0094                   1  2021-12-20  2021   12   20  01:10:00   
714  2021-0095                   1  2021-12-30  2021   12   30  00:43:00   
715  2021-0096                   1  2021-12-15  2021   12   15  10:30:00   
716  2021-0097                   1  2021-11-18  2021   11   18  06:10:00   

     Hora entera Tipo de Calle  Comuna      Pos x      Pos y   Participantes  \
0              4       AVENIDA       8 -58.475340 -34.687570       MOTO-AUTO   
1              1      GRAL PAZ       9 -58.508775 -34.669777  AUTO-PASAJEROS   
2              7       AVENIDA       1 -58.390403 -34.631894       MOTO-AUTO   
3              0       AVENIDA       8 -58.465039 -34.680930         MOTO-SD   
4              5       AVENIDA       1 -58.387183 -34.622466  MOTO-PASAJEROS   
..           ...           ...     ...        ...        ...             ...   
712           17       AVENIDA       7 -58.433538 -34.645616       MOTO-AUTO   
713            1     AUTOPISTA       9 -58.467398 -34.651178       MOTO-AUTO   
714            0       AVENIDA      11 -58.472934 -34.619847     MOTO-CARGAS   
715           10       AVENIDA       9 -58.470668 -34.650217     AUTO-CARGAS   
716            6         CALLE       1 -58.379762 -34.586796  BICICLETA-AUTO   

       Acusado                   Rol    Víctima       Sexo  Edad  
0         AUTO             CONDUCTOR       MOTO  MASCULINO    19  
1    PASAJEROS             CONDUCTOR       AUTO  MASCULINO    70  
2         AUTO             CONDUCTOR       MOTO  MASCULINO    30  
3           SD             CONDUCTOR       MOTO  MASCULINO    18  
4    PASAJEROS             CONDUCTOR       MOTO  MASCULINO    29  
..         ...                   ...        ...        ...   ...  
712       AUTO  PASAJERO_ACOMPAÑANTE       MOTO   FEMENINO    18  
713       AUTO  PASAJERO_ACOMPAÑANTE       MOTO   FEMENINO    43  
714     CARGAS             CONDUCTOR       MOTO  MASCULINO    27  
715     CARGAS             CONDUCTOR       AUTO  MASCULINO    60  
716       AUTO              CICLISTA  BICICLETA  MASCULINO    53  

[717 rows x 18 columns]

In [166]:
#verificamos la informacion general
homicidos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  717 non-null    object 
 1   Numero de víctimas  717 non-null    int64  
 2   Fecha               717 non-null    object 
 3   Año                 717 non-null    int64  
 4   Mes                 717 non-null    int64  
 5   Día                 717 non-null    int64  
 6   Hora                717 non-null    object 
 7   Hora entera         717 non-null    int64  
 8   Tipo de Calle       717 non-null    object 
 9   Comuna              717 non-null    int64  
 10  Pos x               717 non-null    float64
 11  Pos y               717 non-null    float64
 12  Participantes       717 non-null    object 
 13  Acusado             717 non-null    object 
 14  Rol                 717 non-null    object 
 15  Víctima             717 non-null    object 
 16  Sexo    

Exportamos a csv para poder continuar con nuestro trabajo 

In [167]:
homicidos_df.to_csv('C:/Users/CRISTIAN/Desktop/proyecto final # 2/CSV/homicidios_final.csv')

DATOS DE LA POBLACION DE CABA

Según datos del censo de 2022 publicados por el gobierno argentino, la Ciudad Autónoma de Buenos Aires (CABA) tiene una población de 3,121,707 habitantes. Sin embargo, un dato destacado proporcionado por el Consejo Económico y Social de la Ciudad de Buenos Aires indica que cada día ingresan a la ciudad aproximadamente 3.5 millones de personas provenientes de varios municipios del Gran Buenos Aires. Esto significa que la cantidad de personas presentes en la ciudad en un día típico más que duplica su población permanente. Por lo tanto, utilizar únicamente los datos del censo para calcular indicadores clave de rendimiento (KPIs) podría resultar inadecuado y no reflejaría la realidad de la ciudad.

KPI'S

Se nos ha pedido calcular los siguientes KPI:

  1. Reducir en un 10% la tasa de homicidios en siniestros viales de los últimos seis meses, en CABA, en comparación con la tasa de homicidios en siniestros viales del semestre anterior.

Definimos a la tasa de homicidios en siniestros viales como el número de víctimas fatales en accidentes de tránsito por cada 100,000 habitantes en un área geográfica durante un período de tiempo específico. Su fórmula es: (Número de homicidios en siniestros viales / Población total) * 100,000

  2. Reducir en un 7% la cantidad de accidentes mortales de motociclistas en el último año, en CABA, respecto al año anterior.

Definimos a la cantidad de accidentes mortales de motociclistas en siniestros viales como el número absoluto de accidentes fatales en los que estuvieron involucradas víctimas que viajaban en moto en un determinado periodo temporal. Su fórmula para medir la evolución de los accidentes mortales con víctimas en moto es: (Número de accidentes mortales con víctimas en moto en el año anterior - Número de accidentes mortales con víctimas en moto en el año actual) / (Número de accidentes mortales con víctimas en moto en el año anterior) * 100

In [ ]:
#A partir de la base de datos de población y la de homicidios voy a extraer la información necesaria para hacer los calculos correspondientes

Para calcular el primer KPI necesitamos obtener la cantidad de muertes correspondientes de cada semestre de los años registrados, del dataframe 'homicidios_df'

In [168]:
#creamos la columna semestre para hacer menos complicado el calculo
def determinar_semestre(mes):
    if 1 <= mes <= 6:
        return 1
    elif 7 <= mes <= 12:
        return 2
    else:
        return 'mes no valido'

#aplicamos la funcion a la columna 'meses' y creamos una nueva  columna 'semestre'
homicidos_df['Semestre'] = homicidos_df['Mes'].apply(determinar_semestre)



In [169]:
#verificamos
homicidos_df

ID  Numero de víctimas       Fecha   Año  Mes  Día      Hora  \
0    2016-0001                   1  2016-01-01  2016    1    1  04:00:00   
1    2016-0002                   1  2016-01-02  2016    1    2  01:15:00   
2    2016-0003                   1  2016-01-03  2016    1    3  07:00:00   
3    2016-0004                   1  2016-01-10  2016    1   10  00:00:00   
4    2016-0005                   1  2016-01-21  2016    1   21  05:20:00   
..         ...                 ...         ...   ...  ...  ...       ...   
712  2021-0093                   1  2021-12-13  2021   12   13  17:10:00   
713  2021-0094                   1  2021-12-20  2021   12   20  01:10:00   
714  2021-0095                   1  2021-12-30  2021   12   30  00:43:00   
715  2021-0096                   1  2021-12-15  2021   12   15  10:30:00   
716  2021-0097                   1  2021-11-18  2021   11   18  06:10:00   

     Hora entera Tipo de Calle  Comuna      Pos x      Pos y   Participantes  \
0              4       AVENIDA       8 -58.475340 -34.687570       MOTO-AUTO   
1              1      GRAL PAZ       9 -58.508775 -34.669777  AUTO-PASAJEROS   
2              7       AVENIDA       1 -58.390403 -34.631894       MOTO-AUTO   
3              0       AVENIDA       8 -58.465039 -34.680930         MOTO-SD   
4              5       AVENIDA       1 -58.387183 -34.622466  MOTO-PASAJEROS   
..           ...           ...     ...        ...        ...             ...   
712           17       AVENIDA       7 -58.433538 -34.645616       MOTO-AUTO   
713            1     AUTOPISTA       9 -58.467398 -34.651178       MOTO-AUTO   
714            0       AVENIDA      11 -58.472934 -34.619847     MOTO-CARGAS   
715           10       AVENIDA       9 -58.470668 -34.650217     AUTO-CARGAS   
716            6         CALLE       1 -58.379762 -34.586796  BICICLETA-AUTO   

       Acusado                   Rol    Víctima       Sexo  Edad  Semestre  
0         AUTO             CONDUCTOR       MOTO  MASCULINO    19         1  
1    PASAJEROS             CONDUCTOR       AUTO  MASCULINO    70         1  
2         AUTO             CONDUCTOR       MOTO  MASCULINO    30         1  
3           SD             CONDUCTOR       MOTO  MASCULINO    18         1  
4    PASAJEROS             CONDUCTOR       MOTO  MASCULINO    29         1  
..         ...                   ...        ...        ...   ...       ...  
712       AUTO  PASAJERO_ACOMPAÑANTE       MOTO   FEMENINO    18         2  
713       AUTO  PASAJERO_ACOMPAÑANTE       MOTO   FEMENINO    43         2  
714     CARGAS             CONDUCTOR       MOTO  MASCULINO    27         2  
715     CARGAS             CONDUCTOR       AUTO  MASCULINO    60         2  
716       AUTO              CICLISTA  BICICLETA  MASCULINO    53         2  

[717 rows x 19 columns]

para calcular el segundo KPI necesitamos obtener la cantidad de muertes de motocicletas ocurridas, en cada año, de el dataframe 'homicidios_df'

In [170]:
#seleccionamos todos los registros en los que la victima iba en moto
motocicletas_df = homicidos_df[homicidos_df['Víctima'] == 'MOTO']
motocicletas_df

ID  Numero de víctimas       Fecha   Año  Mes  Día      Hora  \
0    2016-0001                   1  2016-01-01  2016    1    1  04:00:00   
2    2016-0003                   1  2016-01-03  2016    1    3  07:00:00   
3    2016-0004                   1  2016-01-10  2016    1   10  00:00:00   
4    2016-0005                   1  2016-01-21  2016    1   21  05:20:00   
5    2016-0008                   1  2016-01-24  2016    1   24  18:30:00   
..         ...                 ...         ...   ...  ...  ...       ...   
708  2021-0089                   1  2021-12-02  2021   12    2  01:10:00   
710  2021-0091                   1  2021-12-11  2021   12   11  23:00:00   
712  2021-0093                   1  2021-12-13  2021   12   13  17:10:00   
713  2021-0094                   1  2021-12-20  2021   12   20  01:10:00   
714  2021-0095                   1  2021-12-30  2021   12   30  00:43:00   

     Hora entera Tipo de Calle  Comuna      Pos x      Pos y  \
0              4       AVENIDA       8 -58.475340 -34.687570   
2              7       AVENIDA       1 -58.390403 -34.631894   
3              0       AVENIDA       8 -58.465039 -34.680930   
4              5       AVENIDA       1 -58.387183 -34.622466   
5             18       AVENIDA       8 -58.444513 -34.684759   
..           ...           ...     ...        ...        ...   
708            1       AVENIDA      11 -58.476337 -34.621406   
710           23         CALLE      10 -58.519894 -34.622849   
712           17       AVENIDA       7 -58.433538 -34.645616   
713            1     AUTOPISTA       9 -58.467398 -34.651178   
714            0       AVENIDA      11 -58.472934 -34.619847   

        Participantes      Acusado                   Rol Víctima       Sexo  \
0           MOTO-AUTO         AUTO             CONDUCTOR    MOTO  MASCULINO   
2           MOTO-AUTO         AUTO             CONDUCTOR    MOTO  MASCULINO   
3             MOTO-SD           SD             CONDUCTOR    MOTO  MASCULINO   
4      MOTO-PASAJEROS    PASAJEROS             CONDUCTOR    MOTO  MASCULINO   
5    MOTO-OBJETO FIJO  OBJETO FIJO             CONDUCTOR    MOTO  MASCULINO   
..                ...          ...                   ...     ...        ...   
708         MOTO-AUTO         AUTO             CONDUCTOR    MOTO  MASCULINO   
710         MOTO-AUTO         AUTO             CONDUCTOR    MOTO  MASCULINO   
712         MOTO-AUTO         AUTO  PASAJERO_ACOMPAÑANTE    MOTO   FEMENINO   
713         MOTO-AUTO         AUTO  PASAJERO_ACOMPAÑANTE    MOTO   FEMENINO   
714       MOTO-CARGAS       CARGAS             CONDUCTOR    MOTO  MASCULINO   

     Edad  Semestre  
0      19         1  
2      30         1  
3      18         1  
4      29         1  
5      30         1  
..    ...       ...  
708    41         2  
710    24         2  
712    18         2  
713    43         2  
714    27         2  

[303 rows x 19 columns]

In [171]:
#agrupamos por el año y obtenemos el total de los accidentes anuales en motos
accidentes_motos_df = motocicletas_df.groupby('Año')['Víctima'].count().reset_index()
accidentes_motos_df

Año  Víctima
0  2016       65
1  2017       56
2  2018       57
3  2019       50
4  2020       29
5  2021       46

In [172]:
accidentes_motos_df.to_csv('C:/Users/CRISTIAN/Desktop/proyecto final # 2/CSV/homicidios_moto.csv')